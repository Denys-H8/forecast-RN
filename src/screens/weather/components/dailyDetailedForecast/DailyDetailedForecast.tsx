import React, { useContext } from 'react';
import { View } from 'react-native';

import HourWeather from '../atomic/hourWeather';
import { styles } from './dailyDetailedForecast.styles';
import { timestampToHour, timestampToDate, getWeatherIconName } from 'src/helpers';
import { useSelector } from 'react-redux';
import { weatherSelector, weatherStatsSelector } from 'src/store/weather/selectors';

type Props = {
  date: string;
};

const DailyForecast: React.FC<Props> = ({ date }) => {
  const weatherStats = useSelector(weatherStatsSelector);
  const weather = useSelector(weatherSelector);

  const hourlyWeather =
    date === 'today'
      ? weatherStats.hourly.slice(0, 5)
      : weather.list
          .filter((el: { dt: number }) => {
            return timestampToDate(el.dt) === date;
          })
          .slice(0, 5);

  const getTemp = (el: { dt: number; temp: number; main: { temp: number }; weather: [{ description: string }] }) => {
    return date === 'today' ? el.temp.toFixed(1) : el.main.temp.toFixed(1);
  };

  return (
    <View style={styles.container}>
      {hourlyWeather.map(
        (el: { dt: number; temp: number; main: { temp: number }; weather: [{ description: string }] }, i: number) => {
          return (
            <HourWeather
              key={el.dt}
              hour={timestampToHour(el.dt)}
              name={getWeatherIconName(el.weather[0].description)}
              temperature={getTemp(el)}
            />
          );
        },
      )}
    </View>
  );
};

export default DailyForecast;
