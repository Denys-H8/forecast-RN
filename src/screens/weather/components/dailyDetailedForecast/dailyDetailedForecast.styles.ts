import { StyleSheet } from 'react-native';

import constants from 'src/constants';

export const styles = StyleSheet.create({
  container: {
    height: 160,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 32,
    borderTopWidth: 2,
    borderBottomWidth: 2,
    borderColor: constants.SUB_COLOR,
  },
});
