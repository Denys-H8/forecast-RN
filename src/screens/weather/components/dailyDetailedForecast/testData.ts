export const weather = {
  list: [
    {
      dt: 1619114400,
      main: {
        temp: 11.84,
        feels_like: 10.7,
        temp_min: 11.84,
        temp_max: 12,
        pressure: 1019,
        sea_level: 1019,
        grnd_level: 1002,
        humidity: 62,
        temp_kf: -0.16,
      },
      weather: [
        {
          id: 800,
          main: 'Clear',
          description: 'clear sky',
          icon: '01n',
        },
      ],
      clouds: {
        all: 0,
      },
      wind: {
        speed: 2.56,
        deg: 195,
        gust: 6.42,
      },
      visibility: 10000,
      pop: 0,
      sys: {
        pod: 'n',
      },
      dt_txt: '2021-04-22 18:00:00',
    },
    {
      dt: 1619125200,
      main: {
        temp: 11.57,
        feels_like: 10.43,
        temp_min: 11.02,
        temp_max: 11.57,
        pressure: 1019,
        sea_level: 1019,
        grnd_level: 1001,
        humidity: 63,
        temp_kf: 0.55,
      },
      weather: [
        {
          id: 802,
          main: 'Clouds',
          description: 'scattered clouds',
          icon: '03n',
        },
      ],
      clouds: {
        all: 28,
      },
      wind: {
        speed: 2.36,
        deg: 185,
        gust: 6.61,
      },
      visibility: 10000,
      pop: 0,
      sys: {
        pod: 'n',
      },
      dt_txt: '2021-04-22 21:00:00',
    },
  ],
  city: {
    id: 706483,
    name: 'Kharkiv',
    coord: {
      lat: 50,
      lon: 36.25,
    },
    country: 'UA',
    population: 1430885,
    timezone: 10800,
    sunrise: 1619058476,
    sunset: 1619109522,
  },
};

export { weatherStatsValue } from 'src/screens/forecast/components/dailyWeather/testValue';
