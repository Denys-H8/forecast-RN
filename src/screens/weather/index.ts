import WeatherScreen from './Weather';

export default WeatherScreen;
