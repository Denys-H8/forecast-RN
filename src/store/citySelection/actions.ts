import * as ACTION from './actionTypes';
import { TGroupWeather, TData, TCity } from './reducer';

export const queryChanged = (query: string, data: TData[]) => ({
  type: ACTION.QUERY_CHANGED,
  payload: { query, data },
});
export const getWeatherForCities = () => ({ type: ACTION.GET_WEATHER_FOR_CITIES });
export const getWeatherForCitiesSucess = (data: TGroupWeather) => ({
  type: ACTION.GET_WEATHER_FOR_CITIES_SUCESS,
  payload: data,
});
export const resetQuery = () => ({ type: ACTION.RESET_QUERY });
export const changeCity = (city: TCity) => ({
  type: ACTION.CHANGE_CITY,
  payload: city,
});
export const pressedCity = (city: TCity) => ({
  type: ACTION.PRESSED_CITY,
  payload: city,
});
export const setLoading = (value: boolean) => ({ type: ACTION.SET_LOADING, payload: value });
