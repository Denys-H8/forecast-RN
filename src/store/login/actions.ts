import * as ACTION from './actionTypes';
import { TUser } from './reducer';

export const login = () => ({ type: ACTION.LOGIN });
export const logout = () => ({ type: ACTION.LOGOUT });
export const logoutSuccess = () => ({ type: ACTION.LOGOUT_SUCCESS });
export const loginStarted = () => ({ type: ACTION.LOGIN_STARTED });
export const loginSuccess = (user: TUser) => ({ type: ACTION.LOGIN_SUCCESS, payload: user });
export const changedEmail = (email: string | false) => ({ type: ACTION.CHANGED_EMAIL, payload: email });
export const changedPassword = (password: string) => ({ type: ACTION.CHANGED_PASSWORD, payload: password });
export const failedEmail = () => ({ type: ACTION.FAILED_EMAIL });
export const failedPassword = () => ({ type: ACTION.FAILED_PASSWORD });
export const startLoading = () => ({ type: ACTION.START_LOADING });
export const loading = (value: boolean) => ({ type: ACTION.LOADING_LOGIN, payload: value });
