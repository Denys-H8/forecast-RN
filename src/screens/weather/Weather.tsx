import React, { useEffect } from 'react';
import { View, ImageBackground } from 'react-native';

import MetaInfo from './components/metaInfo';
import DailyForecast from './components/dailyForecast';
import DailyDetailedForecast from './components/dailyDetailedForecast';
import DayStats from './components/dayStats';
import { styles } from './weather.styles';
import { useDispatch, useSelector } from 'react-redux';
import * as WEATHER_SELECTORS from '../../store/weather/selectors';
import Loading from 'src/components/loading';
import { weatherActions } from 'src/store/rootActions';
import { rootNavigation } from 'src/services';

interface IProps {
  route: { params: { date: string } };
}

const WeatherScreen: React.FC<IProps> = ({ route }: IProps) => {
  const loading = useSelector(WEATHER_SELECTORS.loading);
  const dispatch = useDispatch();
  const lateNavigation = useSelector(WEATHER_SELECTORS.lateNavigationSelector);

  useEffect(() => {
    if (lateNavigation) {
      rootNavigation.navigate(lateNavigation, {});
      dispatch(weatherActions.resetLateNavigation());
    }
  }, []);

  if (loading) {
    return <Loading />;
  }

  return (
    <View style={styles.container}>
      <View style={styles.content}>
        <View>
          <ImageBackground source={require('src/assets/images/cloud.png')} style={styles.image}>
            <MetaInfo date={route.params.date} />
            <DailyForecast date={route.params.date} />
          </ImageBackground>
        </View>
        <DailyDetailedForecast date={route.params.date} />
        <DayStats date={route.params.date} />
      </View>
    </View>
  );
};

export default WeatherScreen;
