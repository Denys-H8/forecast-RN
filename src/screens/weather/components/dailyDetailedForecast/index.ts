import DailyDetailedForecast from './DailyDetailedForecast';

export default DailyDetailedForecast;
