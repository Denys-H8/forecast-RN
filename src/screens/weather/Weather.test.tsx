import React from 'react';
import { render } from '@testing-library/react-native';
import Weather from './Weather';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { weather, weatherStatsValue } from './components/dailyDetailedForecast/testData';
import { resetLateNavigation } from 'src/store/weather/actions';

const mockStore = configureStore([]);

const store = mockStore({
  weatherReducer: {
    loading: false,
    lateNavigation: null,
    weather,
    weatherStats: weatherStatsValue,
  },
});

describe('Weather ', () => {
  test('should render correctly with today', () => {
    const date = 'today';
    Date.now = jest.fn(() => 1619098295 * 1000);
    const { toJSON } = render(
      <Provider store={store}>
        <Weather route={{ params: { date } }} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render correctly with lateNavigation', () => {
    const store = mockStore({
      weatherReducer: {
        loading: false,
        lateNavigation: 'Settings',
        weather,
        weatherStats: weatherStatsValue,
      },
    });
    const date = 'today';

    store.dispatch = jest.fn();

    const { toJSON } = render(
      <Provider store={store}>
        <Weather route={{ params: { date } }} />
      </Provider>,
    );

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(resetLateNavigation());
  });

  test('should render correctly with lateNavigation', () => {
    const store = mockStore({
      weatherReducer: {
        loading: true,
        lateNavigation: 'Settings',
        weather,
        weatherStats: weatherStatsValue,
      },
    });
    const date = 'today';

    const { toJSON } = render(
      <Provider store={store}>
        <Weather route={{ params: { date } }} />
      </Provider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
