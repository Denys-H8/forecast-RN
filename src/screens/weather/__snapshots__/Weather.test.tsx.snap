// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather  should render correctly with lateNavigation 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#191C39",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#fff"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Weather  should render correctly with today 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#191C39",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "marginBottom": 45,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/images/cloud.png",
            }
          }
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": undefined,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "marginTop": 15,
              "paddingLeft": 32,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-ExtraBold",
                  "fontSize": 52,
                  "lineHeight": 61,
                },
                Object {
                  "color": "#fff",
                },
              ]
            }
          >
            Kharkiv
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-Regular",
                },
                Object {
                  "color": "#8B8EA5",
                  "fontSize": 20,
                  "lineHeight": 23.5,
                },
              ]
            }
          >
            THURSDAY, APRIL 22
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 70,
              "justifyContent": "space-between",
              "marginTop": 22,
              "paddingHorizontal": 32,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 60,
                    "lineHeight": 70,
                  },
                ]
              }
            >
              18
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 25,
                    "lineHeight": 30,
                  },
                ]
              }
            >
              0
            </Text>
          </View>
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#8B8EA5",
                  "fontSize": 45,
                },
                undefined,
                Object {
                  "fontFamily": "Fontisto",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          <View>
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#8B8EA5",
                    "fontSize": 12,
                  },
                ]
              }
            >
              SCATTERED
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#8B8EA5",
                    "fontSize": 12,
                  },
                ]
              }
            >
              CLOUDS
            </Text>
          </View>
          <View>
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 18,
                    "lineHeight": 20,
                  },
                ]
              }
            >
              6 /
            </Text>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "WorkSans-Regular",
                    },
                    Object {
                      "color": "#fff",
                      "fontSize": 18,
                      "lineHeight": 20,
                    },
                  ]
                }
              >
                18
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "WorkSans-Regular",
                    },
                    Object {
                      "color": "#fff",
                      "fontSize": 12,
                      "lineHeight": 12,
                    },
                  ]
                }
              >
                0
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "WorkSans-Regular",
                    },
                    Object {
                      "color": "#fff",
                      "fontSize": 18,
                      "lineHeight": 20,
                    },
                  ]
                }
              >
                C
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 2,
          "borderColor": "#8B8EA5",
          "borderTopWidth": 2,
          "flexDirection": "row",
          "height": 160,
          "justifyContent": "space-between",
          "paddingHorizontal": 32,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 90,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#8B8EA5",
                "fontSize": 12,
              },
            ]
          }
        >
          16:00
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 28,
              },
              undefined,
              Object {
                "fontFamily": "Fontisto",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-Regular",
                },
                Object {
                  "color": "#fff",
                  "fontSize": 20,
                },
              ]
            }
          >
            17.3
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-Regular",
                },
                Object {
                  "color": "#fff",
                  "fontSize": 12,
                  "lineHeight": 12,
                },
              ]
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 90,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#8B8EA5",
                "fontSize": 12,
              },
            ]
          }
        >
          17:00
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 28,
              },
              undefined,
              Object {
                "fontFamily": "Fontisto",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-Regular",
                },
                Object {
                  "color": "#fff",
                  "fontSize": 20,
                },
              ]
            }
          >
            17.9
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-Regular",
                },
                Object {
                  "color": "#fff",
                  "fontSize": 12,
                  "lineHeight": 12,
                },
              ]
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 148,
          "justifyContent": "space-between",
          "paddingHorizontal": 32,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <IconMock
          fill="#fff"
          height={32}
          testID="Stat.Icon"
          width={32}
        />
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#fff",
                "fontSize": 20,
                "marginTop": 10,
              },
            ]
          }
        >
          32%
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <IconMock
          fill="#fff"
          height={26}
          testID="Stat.Icon"
          width={26}
        />
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#fff",
                "fontSize": 20,
                "marginTop": 10,
              },
            ]
          }
        >
          6m/c
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <IconMock
          fill="#fff"
          height={28}
          testID="Stat.Icon"
          width={28}
        />
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#fff",
                "fontSize": 20,
                "marginTop": 10,
              },
            ]
          }
        >
          05:27
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <IconMock
          fill="#fff"
          height={28}
          testID="Stat.Icon"
          width={28}
        />
        <Text
          style={
            Array [
              Object {
                "fontFamily": "WorkSans-Regular",
              },
              Object {
                "color": "#fff",
                "fontSize": 20,
                "marginTop": 10,
              },
            ]
          }
        >
          19:38
        </Text>
      </View>
    </View>
  </View>
</View>
`;
